doc/monads/readers.lean

```diff
@@ -0,0 +1,199 @@
+which requires the `IO` monad.
+-/
+
+structure Environment where
+  path : String
+  home : String
+  user : String
+  deriving Repr
+
+def getEnvDefault (name : String): IO String := do
+  let val? ← IO.getEnv name
+  pure <| match val? with
+          | none => ""
+          | some s => s
+
+def loadEnv : IO Environment := do
+  let path ← getEnvDefault "PATH"
+  let home ← getEnvDefault "HOME"
+  let user ← getEnvDefault "USER"
+  pure { path, home, user }
+
+def func1 (e : Environment) : Float :=
+  let l1 := e.path.length
+  let l2 := e.home.length * 2
+  let l3 := e.user.length * 3
+  (l1 + l2 + l3).toFloat * 2.1
+
+def func2 (env : Environment) : Nat :=
+  2 + (func1 env).floor.toUInt32.toNat
+
+def func3 (env : Environment) : String :=
+  "Result: " ++ (toString (func2 env))
+
+def main : IO Unit := do
+  let env ← loadEnv
+  let str := func3 env
+  IO.println str
+
+#eval main -- Result: 7538
+
+/-!
+The only function actually using the environment is func1. However func1 is a pure function. This
+means it cannot directly call loadEnv, an impure function in the IO monad. This means the
+environment has to be passed through as a variable to the other functions, just so they can
+ultimately pass it to func1. In a language with global variables, you could save env as a global
+value in main. Then func1 could access it directly. There would be no need to have it as a parameter
+to func1, func2 and func3. In larger programs, these "pass-through" variables can cause a lot of
+headaches.
+
+## The Reader Solution
+
+The `ReaderM` monad solves this problem. It effectively creates a global read-only value of a
+specified type. All functions within the monad can "read" the type. Let's look at how the `ReaderM`
+monad changes the shape of this code. Now the functions **no longer need** to be given the
+`Environment` as an explicit parameter, as they can access it through the monad.
+-/
+
+def readerFunc1 : ReaderM Environment Float := do
+  let env ← read
+  let l1 := env.path.length
+  let l2 := env.home.length * 2
+  let l3 := env.user.length * 3
+  return (l1 + l2 + l3).toFloat * 2.1
+
+def readerFunc2 : ReaderM Environment Nat :=
+  readerFunc1 >>= (fun x => return 2 + (x.floor.toUInt32.toNat))
+
+def readerFunc3 : ReaderM Environment String := do
+  let x ← readerFunc2
+  return "Result: " ++ (toString x)
+
+def main2 : IO Unit := do
+  let env ← loadEnv
+  let str := readerFunc3 env
+  IO.println str
+
+#eval main2 -- Result: 7538
+/-!
+`main2` loads the environment as before, and estabilishes the `ReaderM` context by passing `env` to
+readerFunc3.  The `do` notation you learned about in [Monads](monads.lean.md) is hiding some things
+here.  Technically `readerFunc3` is a monadic action and in order to "run" that action the `ReaderM`
+monad provides a `run` method and it is the `ReaderM` run method that takes the initial
+`Environment` context.  So you can actually write the complete code by doing this: `let str :=
+readerFunc3.run env`.
```

Comment:
I think I would like it better if `readerFunc3.run env` was presented as the "canonical" code, and then later in the `ReaderM` implementation section you can state that this is equivalent to `readerFunc3 env` (unless we enact https://github.com/leanprover/lean4/issues/1547#issuecomment-1231838157 :) ).

Reply:
Fixed in https://github.com/lovettchris/lean4/commit/aa8a2586f27ea31f3c8d4a32893322ed5536d43f